tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/EagerOperation.java

```diff
@@ -131,43 +145,104 @@ public Shape shape(int outputIndex) {
   private static class NativeReference extends EagerSession.NativeReference {
 
     NativeReference(
-        EagerSession session, EagerOperation operation, long opHandle, long[] outputHandles) {
+        EagerSession session, EagerOperation operation, TFE_Op opHandle, TFE_TensorHandle[] outputHandles) {
       super(session, operation);
       this.opHandle = opHandle;
       this.outputHandles = outputHandles;
     }
 
     @Override
     void delete() {
-      if (opHandle != 0L) {
+      if (opHandle != null && !opHandle.isNull()) {
         for (int i = 0; i < outputHandles.length; ++i) {
-          if (outputHandles[i] != 0L) {
+          if (outputHandles[i] != null && !outputHandles[i].isNull()) {
             EagerOperation.deleteTensorHandle(outputHandles[i]);
-            outputHandles[i] = 0L;
+            outputHandles[i] = null;
           }
         }
         EagerOperation.delete(opHandle);
-        opHandle = 0L;
+        opHandle = null;
       }
     }
 
-    private long opHandle;
-    private final long[] outputHandles;
+    private TFE_Op opHandle;
+    private final TFE_TensorHandle[] outputHandles;
   }
-  
-  private static native void delete(long handle);
 
-  private static native void deleteTensorHandle(long handle);
+  private static void requireOp(TFE_Op handle) {
+    if (handle == null || handle.isNull()) {
+      throw new IllegalStateException("Eager session has been closed");
+    }
+  }
 
-  private static native long resolveTensorHandle(long handle);
+  private static void requireTensorHandle(TFE_TensorHandle handle) {
+    if (handle == null || handle.isNull()) {
+      throw new IllegalStateException("EagerSession has been closed");
+    }
+  }
 
-  private static native int outputListLength(long handle, String name);
+  private static void delete(TFE_Op handle) {
+    if (handle == null || handle.isNull()) return;
+    TFE_DeleteOp(handle);
+  }
 
-  private static native int inputListLength(long handle, String name);
+  private static void deleteTensorHandle(TFE_TensorHandle handle) {
+    if (handle == null || handle.isNull()) return;
+    TFE_DeleteTensorHandle(handle);
+  }
 
-  private static native int dataType(long handle);
+  private static TF_Tensor resolveTensorHandle(TFE_TensorHandle handle) {
+    requireTensorHandle(handle);
+    try (PointerScope scope = new PointerScope()) {
+      TF_Status status = TF_Status.newStatus();
+      TF_Tensor tensor = TFE_TensorHandleResolve(handle, status);
+      status.throwExceptionIfNotOK();
+      return tensor;
+    }
+  }
 
-  private static native int numDims(long handle);
+  private static int outputListLength(TFE_Op handle, String name) {
+    requireOp(handle);
+    try (PointerScope scope = new PointerScope()) {
+      TF_Status status = TF_Status.newStatus();
+      int length = TFE_OpGetOutputLength(handle, name, status);
+      status.throwExceptionIfNotOK();
+      return length;
+    }
```

Comment:
Instead of starting new scopes in all of these methods, could it be simpler and more efficient to just create the status in a `try-with-resource` block when there are not other resource allocated?
```java
try (TF_Status status = TF_Status.newStatus()) { 
    ... 
}
```

Reply:
Yes, it would be more efficient, but it would also make it more error prone when we start creating other objects in there that may start doing temporary allocations.

Reply:
Yeah... I would still prefer we go with the most efficient approach but it's up to you if you want to make the changes or not, we can merge it like this too.